File: shuttle-api/src/ShuttleApi.Domain/Drivers/DriverDocument.cs
using ShuttleApi.Domain.Common;

namespace ShuttleApi.Domain.Drivers;

public sealed class DriverDocument : Entity<Guid>
{
    public Guid DriverId { get; private set; }
    public DocumentType DocumentType { get; private set; }
    public string FileName { get; private set; } = string.Empty;
    public string ContentType { get; private set; } = string.Empty;

    /// <summary>
    /// Opaque reference to the stored file. For the DB implementation this is a GUID string;
    /// for future blob storage this will be a blob URL/path.
    /// </summary>
    public string StorageKey { get; private set; } = string.Empty;

    public long FileSizeBytes { get; private set; }
    public DateTime UploadedAt { get; private set; }
    public DateTime? ExpiryDate { get; private set; }

    // Drug & Alcohol Test fields
    public DateTime? TestDate { get; private set; }
    public TestResult? TestResultValue { get; private set; }
    public string? TestedBy { get; private set; }

    // Driver's License fields
    public string? LicenseNumber { get; private set; }
    public LicenseClass? LicenseClass { get; private set; }
    public DateTime? IssuedDate { get; private set; }
    public string? LicenseProvince { get; private set; }

    // Police Record Check fields
    public CheckResult? CheckResultValue { get; private set; }
    public string? IssuingAuthority { get; private set; }

    // Driver Abstract fields
    public int? ViolationCount { get; private set; }
    public int? AtFaultAccidentCount { get; private set; }

    // Shared nullable field
    public string? Notes { get; private set; }

    public bool IsExpiringSoon =>
        ExpiryDate.HasValue && ExpiryDate.Value <= DateTime.UtcNow.AddDays(60);

    private DriverDocument() { }

    public static DriverDocument Create(
        Guid driverId,
        DocumentType documentType,
        string fileName,
        string contentType,
        string storageKey,
        long fileSizeBytes,
        DateTime? expiryDate,
        DateTime? testDate,
        TestResult? testResultValue,
        string? testedBy,
        string? licenseNumber,
        LicenseClass? licenseClass,
        DateTime? issuedDate,
        string? licenseProvince,
        CheckResult? checkResultValue,
        string? issuingAuthority,
        int? violationCount,
        int? atFaultAccidentCount,
        string? notes)
    {
        return new DriverDocument
        {
            Id = Guid.NewGuid(),
            DriverId = driverId,
            DocumentType = documentType,
            FileName = fileName,
            ContentType = contentType,
            StorageKey = storageKey,
            FileSizeBytes = fileSizeBytes,
            UploadedAt = DateTime.UtcNow,
            ExpiryDate = expiryDate,
            TestDate = testDate,
            TestResultValue = testResultValue,
            TestedBy = testedBy,
            LicenseNumber = licenseNumber,
            LicenseClass = licenseClass,
            IssuedDate = issuedDate,
            LicenseProvince = licenseProvince,
            CheckResultValue = checkResultValue,
            IssuingAuthority = issuingAuthority,
            ViolationCount = violationCount,
            AtFaultAccidentCount = atFaultAccidentCount,
            Notes = notes
        };
    }
}

File: shuttle_tablet/lib/features/drivers/domain/entities/driver_document.dart
import 'package:equatable/equatable.dart';

enum DocumentType { drugAndAlcoholTest, driversLicense, policeRecordCheck, driverAbstract }

enum TestResult { pass, fail, pending }

enum CheckResult { clear, conditions, flagged }

enum LicenseClass { a, b, c, d, e, f }

class DriverDocument extends Equatable {
  final String id;
  final String driverId;
  final DocumentType documentType;
  final String fileName;
  final String contentType;
  final int fileSizeBytes;
  final DateTime uploadedAt;
  final DateTime? expiryDate;
  final bool isExpiringSoon;

  // Drug & Alcohol Test
  final DateTime? testDate;
  final TestResult? testResultValue;
  final String? testedBy;

  // Driver's License
  final String? licenseNumber;
  final LicenseClass? licenseClass;
  final DateTime? issuedDate;
  final String? licenseProvince;

  // Police Record Check
  final CheckResult? checkResultValue;
  final String? issuingAuthority;

  // Driver Abstract
  final int? violationCount;
  final int? atFaultAccidentCount;

  final String? notes;

  String get documentTypeLabel => switch (documentType) {
        DocumentType.drugAndAlcoholTest => 'Drug & Alcohol Test',
        DocumentType.driversLicense => "Driver's License",
        DocumentType.policeRecordCheck => 'Police Record Check',
        DocumentType.driverAbstract => 'Driver Abstract',
      };

  const DriverDocument({
    required this.id,
    required this.driverId,
    required this.documentType,
    required this.fileName,
    required this.contentType,
    required this.fileSizeBytes,
    required this.uploadedAt,
    this.expiryDate,
    this.isExpiringSoon = false,
    this.testDate,
    this.testResultValue,
    this.testedBy,
    this.licenseNumber,
    this.licenseClass,
    this.issuedDate,
    this.licenseProvince,
    this.checkResultValue,
    this.issuingAuthority,
    this.violationCount,
    this.atFaultAccidentCount,
    this.notes,
  });

  @override
  List<Object?> get props => [
        id,
        driverId,
        documentType,
        fileName,
        contentType,
        fileSizeBytes,
        uploadedAt,
        expiryDate,
        isExpiringSoon,
        testDate,
        testResultValue,
        testedBy,
        licenseNumber,
        licenseClass,
        issuedDate,
        licenseProvince,
        checkResultValue,
        issuingAuthority,
        violationCount,
        atFaultAccidentCount,
        notes,
      ];
}
